packages/fxa-auth-server/scripts/convert-customers-to-stripe-automatic-tax.ts

```diff
@@ -0,0 +1,328 @@
+const generateReportForSubscription = async (
+  stripe: Stripe,
+  firestoreSubscription: FirestoreSubscription,
+  outputFile: string
+) => {
+  try {
+    const {
+      id: subscriptionId,
+      customer: customerId,
+      plan,
+    } = firestoreSubscription;
+
+    const product = await stripe.products.retrieve(plan.product as string);
+
+    const customer = await fetchCustomer(stripe, customerId);
+    if (!customer) return;
+
+    if (!isTaxEligible(customer)) {
+      // TODO FXA-6575: IP address mapping
+      await enableStripeTax(stripe, customerId, {});
+
+      const updatedCustomer = await fetchCustomer(stripe, customerId);
+
+      // The customer is not in a taxable location
+      if (!updatedCustomer || !isTaxEligible(updatedCustomer)) return;
+    }
+
+    await stripe.subscriptions.update(subscriptionId, {
+      automatic_tax: {
+        enabled: true,
+      },
+    });
+
+    const invoicePreview = await stripe.invoices.retrieveUpcoming({
+      subscription: subscriptionId,
+      expand: ['total_tax_amounts.tax_rate'],
+    });
+
+    const report = buildReport({
+      customer,
+      subscription: firestoreSubscription,
+      product,
+      plan,
+      invoicePreview,
+    });
+
+    // TODO FXA-6581: Write report to CSV
+    console.log('report:', report);
+  } catch (e) {
+    // TODO FXA-6581: Error output formatting
+    console.error(e);
+  }
+};
+
+// Converts all eligible customers to Stripe tax given the following constraints:
+// 1. Customer must exist in the provided IP address mapping
+// 2. Stripe must be able to place the customer in a taxable location
+// 3. Upcoming subscription renewal must be more than 14 days in the future for
+//    1 month subscriptions, and more than 30 days in the future for 6 month and
+//    yearly subscriptions.
+const convertCustomersToStripeTax = async (args: {
+  isDryRun: boolean;
+  batchSize: number;
+  outputFile: string;
+}) => {
+  const config = Container.get(AppConfig);
+  const firestore = Container.get(AuthFirestore);
+  const stripe = new Stripe(config.subscriptions.stripeApiKey, {
+    apiVersion: '2022-08-01',
+    maxNetworkRetries: 3,
+  });
+
+  const collectionPrefix = `${config.authFirestore.prefix}stripe-`;
+  const subscriptionCollection = `${collectionPrefix}subscriptions`;
+
+  let startAfter: string | null = null;
+  let hasMore = true;
+
+  while (hasMore) {
+    const subscriptionSnap = await firestore
+      .collectionGroup(subscriptionCollection)
+      .where('automatic_tax.enabled', '==', false)
+      .orderBy('id')
+      .startAfter(startAfter)
+      .limit(args.batchSize)
+      .get();
+
+    startAfter = subscriptionSnap.docs.at(-1)?.data().id as string;
+    if (!startAfter) hasMore = false;
+
+    const applicableSubs = subscriptionSnap.docs
+      .map((doc) => doc.data() as FirestoreSubscription)
+      .filter(willBeRenewed)
+      .filter(isStripeTaxDisabled)
+      .filter(isWithinNoticePeriod);
```

Comment:
What are the pros/cons to using `.filter().filter().filter()` versus a single `.filter()` w/ `(doc) => willBeRenewed(doc) && isStripeTaxDisabled(doc) && isWithinNoticePeriod(doc)` and only needing to loop once?

Reply:
Readability. The performance implication is minimal since the only additional expense is setting up the loop each time. Filter also produces a shallow copy.

O(3n) is still O(n).